examples/particle-swarm-optimizer.dx

```diff
@@ -0,0 +1,96 @@
+' # Particle Swarm Optimizer
+
+' ## Fitness function
+
+rosenbrock:: Real -> Real -> Real
+rosenbrock x y = sq (1.0 - x) + 10000.0*sq (y - x*x) 
+
+
+' We write one that uses vector for input
+
+rosenbrock2:: (2=>Real) -> Real
+rosenbrock2 xs =
+  x = xs.(asidx 0)
+  y = xs.(asidx 1)
+  rosenbrock x y
+
+
+' Min should be at 1.0, 1.0
+
+:p rosenbrock 1.0 1.000
+
+:p rosenbrock2 [1.0, 1.000]
+
+:p rosenbrock 1.0 1.02
+
+:p rosenbrock2 [1.0, 1.02]
+
+
+
+' ## Helper functions
+
+' make a random vector unifornly distrubuted between lb and ub
+
+randBounded:: Key -> (d=>Real)->(d=>Real)->(d=>Real)
+randBounded key lb ub =
+    for i. lb.i + ((rand $ ixkey key i) * (ub.i - lb.i))
+
+randBounded (newKey 4) [1.0,  -2.0] [-1.0,  2.0]
+
+' minby and minimumby
+' to find the smallest values
+
+minby:: A a::Data. (a->Real)->a->a->a
```

Comment:
I don't know what the `A` is for.

Reply:
It means "for all" (an ASCII approximation to ∀). When you write `f :: a -> a` in Haskell or Dex, there's an implicit "for all a" in front: `f :: ∀ a. a -> a`. Practically, it means that `f` takes a *type* argument, like `Int` or `Real -> Real`, which it binds to `a`. Usually these type arguments are applied implicitly, filled in by type inference, but you can write type application explicitly as `f @Int`. You can also be explicit about the "for all". One reason would be to bind type variables that don't appear in the function's type at all, but only in its body. An example of this is `size` in the prelude:

    size :: A n::Ix. Int
    size = isum for i::n. 1

This is also the way to pass `nparticles` and `niters` in your particle swarm optimizer.

Reply:
Cases like this are really crying out for index literals. There's nothing remotely dynamic about this pair-to-2-vector conversion, so it's a shame to have to use the unchecked `asidx` (which should probably be called `unsafeAsIdx`). [We already have index literals in the abstract syntax](https://github.com/google-research/dex-lang/blob/master/src/lib/Syntax.hs#L56) we just haven't exposed them as concrete syntax. Perhaps we could try something like `0@3`, meaning "index 0 of the index set {0,1,2}". seems reasonable.